src/rdkafka_error.c

```diff
@@ -0,0 +1,190 @@
+
+
+void rd_kafka_error_destroy (rd_kafka_error_t *error) {
+        rd_free(error);
+}
+
+
+/**
+ * @brief Creates a new error object using the optional va-args format list.
+ */
+rd_kafka_error_t *rd_kafka_error_new_v (rd_kafka_resp_err_t code,
+                                        const char *fmt, va_list ap) {
+        rd_kafka_error_t *error;
+        ssize_t strsz = 0;
+
+        if (fmt && *fmt) {
+                va_list ap2;
+                va_copy(ap2, ap);
+                strsz = rd_vsnprintf(NULL, 0, fmt, ap2) + 1;
+                va_end(ap2);
+        }
+
+        error = rd_malloc(sizeof(*error) + strsz);
+        error->code = code;
+        error->fatal = rd_false;
+        error->retriable = rd_false;
+        error->txn_abortable = rd_false;
+
+        if (strsz > 0) {
+                error->errstr = (char *)(error+1);
+                rd_vsnprintf(error->errstr, strsz, fmt, ap);
+        } else {
+                error->errstr = NULL;
+        }
+
+        return error;
+}
+
+rd_kafka_error_t *rd_kafka_error_new (rd_kafka_resp_err_t code,
+                                      const char *fmt, ...) {
+        rd_kafka_error_t *error;
+        va_list ap;
+
+        va_start(ap, fmt);
+        error = rd_kafka_error_new_v(code, fmt, ap);
+        va_end(ap);
+
+        return error;
+}
+
+rd_kafka_error_t *rd_kafka_error_new_fatal (rd_kafka_resp_err_t code,
+                                                const char *fmt, ...) {
+        rd_kafka_error_t *error;
+        va_list ap;
+
+        va_start(ap, fmt);
+        error = rd_kafka_error_new_v(code, fmt, ap);
+        va_end(ap);
+
+        rd_kafka_error_set_fatal(error);
+
+        return error;
+}
+
+rd_kafka_error_t *rd_kafka_error_new_retriable (rd_kafka_resp_err_t code,
+                                                const char *fmt, ...) {
+        rd_kafka_error_t *error;
+        va_list ap;
+
+        va_start(ap, fmt);
+        error = rd_kafka_error_new_v(code, fmt, ap);
+        va_end(ap);
+
+        rd_kafka_error_set_retriable(error);
+
+        return error;
+}
+
+rd_kafka_error_t *rd_kafka_error_new_txn_abortable (rd_kafka_resp_err_t code,
+                                                    const char *fmt, ...) {
+        rd_kafka_error_t *error;
+        va_list ap;
+
+        va_start(ap, fmt);
+        error = rd_kafka_error_new_v(code, fmt, ap);
+        va_end(ap);
+
+        rd_kafka_error_set_txn_abortable(error);
+
+        return error;
+}
+
+
+rd_kafka_resp_err_t rd_kafka_error_code (const rd_kafka_error_t *error) {
+        return error->code;
+}
+
+const char *rd_kafka_error_name (const rd_kafka_error_t *error) {
+        return rd_kafka_err2name(error->code);
+}
+
+const char *rd_kafka_error_string (const rd_kafka_error_t *error) {
+        return error->errstr ? error->errstr : rd_kafka_err2str(error->code);
+}
+
+int rd_kafka_error_is_fatal (const rd_kafka_error_t *error) {
+        return error->fatal ? 1 : 0;
```

Comment:
why the ternary operator?  x3

Reply:
Avoids implicit cast warnings on various compilers.